### ISSUE TITLE
How to: prevent automatic autocomplete popup and only show on keystroke (C-Space)? How to only show diagnostics on save?

### ISSUE THREAD

**Issue by bit-void** [AstroNvim/AstroNvim#1014]
### Question

Hello, I'm struggling to figure out how to disable all visual noise when typing. In general, when I'm writing code I only want autocomplete/diagnostics when I reach for it. In particular, it's distracting to see linter/compiler errors when I'm typing mid-sentence. Is there an easy way to move diagnostics to on save and autocomplete to only show on a keybinding such as ctrl+space? Sorry if this isn't the appropriate spot, I'm loving astrovim and this one tweak would make it perfect. I'm lost in all the layers so not quite sure how to accomplish this, or if it's even possible. Thank you!

**Comment by mehalter** [comment]
Hm, I'm not sure how to do the diagnostic updating only on save. AstroNvim is coming with a new UI/UX set of keybindings that will actually let you toggle the diagnostics on and off on demand. This might be something that could help with this. (The official stable release of AstroNvim 2.0 is October 7 in a little over a week)

For the `cmp` you can disable the autocompletion nature by updating the `cmp` settings. Here is a minimal `user/init.lua` file:

```lua
return {
  plugins = {
    cmp = {
      completion = { autocomplete = false },
    },
  },
}
```

This will make `cmp` not autocomplete and just open if you do `<c-space>`.

**Comment by bit-void** [comment]
@mehalter Ah so much better, thank you! I was having a hard time understanding how to override defaults, that makes sense.

Hmm, I wonder about the diagnostics... I do want the diagnostics just not so quickly and would love them on save, or some other keybinding. The LSP modes require sending the buffer on every keystroke to get completion right? I can't find a way to separate the completion from diagnostics.

**Comment by mehalter** [comment]
Yeah the easiest way to think about the overriding method in AstroNvim is that all of the plugins follow a similar approach to configuration where you set them up with a `require("plugin_name").setup({...})` call. So for `cmp` for example you would do `require("cmp").setup({})`

The `plugins` table in your user configuration allows you to modify the plugin table in the `plugins.init` table and all of the other tables inside would override the table passed to `setup()` function for the specified plugin where the key is the same name as what is used in the `require()` call.

So

```lua
return {
  plugins = {
    cmp = {
      completion = { autocomplete = false },
    },
  },
}
```

This has `plugins.cmp` defined as a table so if will add the options you supplied to `require("cmp").setup({ completion = { autocomplete = false } })`

Also to note, this does merge your options with the default options that we add as well with preference for what the user has defined. So if we had `autocomplete = true` specified in our table, you would override that to be `autocomplete = false`. In this case we aren't specifying it in our table and you are just adding that option to the table.

Sorry this is so much information 😅 hopefully it helps clear some stuff up.

TL;DR `plugins.X = { some_table }` will override `require("X").setup({ some_table })`

**Comment by mehalter** [comment]
If you haven't had a chance yet, this page in the documentation does a better job at explaining this: https://astronvim.github.io/Configuration/override_formats

As well as another override format that we support that gives the user much more control over the configuration override :)

**Comment by bit-void** [comment]
I appreciate all the info thanks!

**Comment by mehalter** [comment]
Just a heads up for anyone coming across this, if you are using AstroNvim v2.0 this is a default option. If you want to disable `cmp` autocompletion on start (`<c-space>` will still open completion menu) and disable `diagnostics` on start, then you can do the following in your `user/init.lua` file:

```lua
return {
  options = {
    g = {
      cmp_enabled = false,
      diagnostics_enabled = false,
    }
  }
}
```

Then `<leader>+ud` will toggle diagnostics and `<leader>+uc` will toggle auto completion.

**Comment by mehalter** [comment]
@bit-void I did find an option that you might be interested in for diagnostics. You can set `update_in_insert` to `false` to disable diagnostics updating while typing! 😄 

So this would look like this in a `user/init.lua` file:

```lua
return {
  diagnostics = {
    update_in_insert = false,
  },
}
```

**Comment by abctaylor** [comment]
I can't get this working for me and it's driving me nuts. What am I doing wrong? The entirety of my `init.lua` looks like this:

```
return {
  options = {
    g = {
      cmp_enabled = false,
      diagnostics_enabled = false,
    }
  }
}
```

The exact file I am editing is `~/.config/nvim/lua/user/init.lua`.

`<space>` `u` `c` turns off completion but it still opens when I open a file.

Thanks.

**Comment by mehalter** [comment]
@abctaylor since v4 this is now configured through AstroCore: https://github.com/AstroNvim/astrocore

**Comment by abctaylor** [comment]
@mehalter I set 
```
features = {
  ...
  cmp = false, 
  ...
```
which I found from the page you linked but keep getting those very annoying auto complete popups for silly things that slow down dev time by 10x. What am I doing wrong? Any help would be awesome.

**Comment by mehalter** [comment]
@abctaylor I cannot replicate this behavior you are describing. If I set `cmp = false` in AstroCore then the automatic completion goes away. Can you share the entire file you are working with so I can see what you are doing wrong.
